### ISSUE TITLE
Statistical significance testing of decoding results

### ISSUE THREAD

**Issue by SylvainTakerkart** [nilearn/nilearn#2483]

**Comment by bthirion** [comment]
Would two-sample permutation tricks do the tricks, where the "control" sample is the chance level ? 
Other wise, is there some relevant literature ?

**Comment by thomasbazeille** [comment]
I think that having some good signifiance testing would be really valuable to offer some fullproof general routines for predictive analysis. However there is no general method available in scikit-learn for common estimators besides permutation testing that take forever, and I'm not aware of any specific litterature on that ? I'll take a look.

 to really worth investigating, however

**Comment by SylvainTakerkart** [comment]
there's a bit of literature, but not much... I'll summarizing what I know of next week...

**Comment by PeerHerholz** [comment]
Hi gang,

sorry for joining late. 
I think there are a couple of approaches/ways we could/should consider:

1) difference between searchlights and ROI decoding
2) significance tests

While permutation tests for searchlights were proposed a while back ([Stelzer et al. 2013](https://www.uwo.ca/bmi/cbs/img/Stelzer_inferenceMVPA_2013.pdf), [pymvpa implementation](http://www.pymvpa.org/generated/mvpa2.algorithms.group_clusterthr.GroupClusterThreshold.html)) there's still no "consistency" I think, as folks are also using parametric tests to evaluate searchlights at the group level. There are also other things, like [prevalence analyses](https://www.sciencedirect.com/science/article/pii/S1053811916303470?casa_token=G360xWJ3KRwAAAAA:LriUtPty_-Mt8yIH5rHXbO9RWLsX6MtplE-X29Q6ElYT0MmeNV3wMVjtjhZVEsqgHjPN3KAdY14). WRT computation time: we could reduce that by different means. E.g. in [fmriflows](https://github.com/miykael/fmriflows/blob/master/notebooks/06_analysis_multivariate.ipynb) we skip a certain amount of voxels and use interpolation later on.
For ROI decoding we could utilize [permutation_test_score](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.permutation_test_score.html), but as @thomasbazeille mentioned, they tend to be very slow and need to be adapted a bit.

Another thingy we could/should discuss are [spatio-temporal analyses](http://www.pymvpa.org/tutorial_eventrelated.html#from-timeseries-to-spatio-temporal-samples).


The significance testing page of the pymvpa docs for reference: [here](http://www.pymvpa.org/tutorial_significance.html#null-hypothesis-testing).

**Comment by banilo** [comment]
First off, one should wonder whether p-values are actually meaningful or
necessary for the type of judgment one wants to make about the data in the
context of searchlight and other decoding analyses. Some of my thoughts:

1) T-test-type approaches of the actual and predicted labels are often
suffering from the typical overlap of the training sets that are used to
train the models that are used to make the predictions. T-testing based on
these overlapping trains sets tend to be spuriously optimistic.
2) Non-parametric permutation tests yield empirical null distributions
(e.g. label switching) that are often better suited to forming p-values for
testing relevant differences between actual and predicted labels
3) Bootstrapping analysis of the whole searchlight/decoding pipeline is
another viable way that may be still under-used (20-25 iterations can be
already informative)
4) Aiming at whole-model out-of-sample predictions and deriving p-values to
isolate single-variable effects correspond to two largely distinct
modelling paradigms:

https://www.frontiersin.org/articles/10.3389/fnins.2017.00543/full

https://www.cell.com/trends/neurosciences/fulltext/S0166-2236(19)30007-4?rss=yes

https://www.nature.com/articles/s42256-019-0069-5?proof=trueMay

Danilo

**Comment by SylvainTakerkart** [comment]
Hi gang,

Sorry, I vanished... (and I'm not going to be of much help during the whole summer...)

My feeling is that we might want to think about starting simple (coz all the questions and comments listed above might be a bit overwhelming to start with), with a basic case of a single-subject ROI decoding. And then we could build up from there by moving to the group-level, and/or jumping into the searchlight realm, to spatio/temporal analyses etc. The length of the pymvpa page is a good clue that nothing is simple...

For the single-subject ROI decoding case, maybe we could provide an example of several approaches that can be used to test the significance of the mean accuracy across the folds of a crossval (permutation of labels using the permutation_test_score of sklearn, t-tests etc.), and demonstrate what's good or not with each of them (computational cost, inflated p-values etc.)?

**Comment by bthirion** [comment]
Thx for the comments.
There are multiple issues indeed:
* t-tests are inaccurate, and I find it somewhat problematic to showcase an approach that we find misleading and should be avoided. 
* the other approaches are computationally prohibitive.

To me, as suggested by  @SylvainTakerkart the best way forward is to select a few positions in an ROI (e.g.  a few tens sphere positions) and run a permutation test.

**Comment by PeerHerholz** [comment]
Sounds about right, thx!
A step-wise procedure would indeed be very helpful.

WRT `ROI` decoding multiple common cases that are present in the literature should be addressed IMHO:

- assess significance within participant 
- assess significance across participants (group level)
- on both above mentioned levels: compare `conditions` and `ROI`s (within `ROI`: is one condition better "decodable" than another, between `ROI`s: is one `ROI` "better" at decoding than another)

**Comment by SylvainTakerkart** [comment]
>     * t-tests are inaccurate, and I find it somewhat problematic to showcase an approach that we find misleading and should be avoided.

I agree!

unless we provide an example that demontrates that the t-test (or other approaches) are misleading... in our recent paper (that deals with the intersubject case, but that's not immediately relevant for this point of the discussion), we've used a simple way to estimate the false positive rate given by a test: using simple artificials dataset where no effect is present, performing decoding, significance testing (with whatever method), and counting the number of false detections...
(see https://doi.org/10.1016/j.neuroimage.2019.116205 , supplementary materials, section 5)

it might be didactic for users, but maybe it's too much??

**Comment by SylvainTakerkart** [comment]
> Sounds about right, thx!
> A step-wise procedure would indeed be very helpful.
> 
> WRT `ROI` decoding multiple common cases that are present in the literature should be addressed IMHO:
> 
>     * assess significance within participant
> 
>     * assess significance across participants (group level)
> 
>     * on both above mentioned levels: compare `conditions` and `ROI`s (within `ROI`: is one condition better "decodable" than another, between `ROI`s: is one `ROI` "better" at decoding than another)

this sounds good to me! except maybe for the last part ("one ROI decodes better than another") which is always tricky...

**Comment by bthirion** [comment]
We might refer to such experiments in the doc, but not include them directly. 
B

**Comment by PeerHerholz** [comment]
I agree @SylvainTakerkart, it's tricky. I mentioned it because I've seen it before in one or the other publication. I guess this goes back to @bthirion's point: which examples should be shown and explained and to what extent? Also should potential/shown problems of other approaches be addressed? Based on the comments above it seems that assessing significance in `ROI` decoding, within and across participants, should be the first step. Should we try to base it on/build upon the [ROI decoding in the Haxby dataset example](https://nilearn.github.io/auto_examples/02_decoding/plot_haxby_full_analysis.html#sphx-glr-auto-examples-02-decoding-plot-haxby-full-analysis-py)?

**Comment by bthirion** [comment]
+1 for drafting something based on this example. Thx !

**Comment by PeerHerholz** [comment]
Hi everyone,

I hope you and your loved ones are still doing ok during these difficult times.
Given that a new release might be close, I was wondering how we should proceed wrt this part of the docs?

Cheers, Peer

**Comment by bthirion** [comment]
@PeerHerholz I would not put release-related pressure on that one, since we need to find some compromises between speed and expressiveness of the statistical framework.

**Comment by SylvainTakerkart** [comment]
Hi!
Let's try to relaunch this discussion... I was wondering whether it could be useful to split our discussions in several use cases... I see two factors that could help structure our work:
- factor 1 = individual vs. group level;
- factor 2 = ROI-based vs. searchlight decoding... 
That would give us four different usecases that we could tackle separately (in terms of method, if separation is needed; but also in terms of priority / timing for implementation)...
WDYT?

**Comment by bthirion** [comment]
And maybe start with the simple case: decoding  / ROI-based decoding.
+ The group-level scenario is probably simpler.

**Comment by SylvainTakerkart** [comment]
> And maybe start with the simple case: decoding / ROI-based decoding.

OK, this makes sense to me! (which does not prevent from discussing what you've recently implemented @PeerHerholz ;) )

>     * The group-level scenario is probably simpler.

can you elaborate on this?

coz I have the opposite intuition, i.e I think that "individual level" might be the easiest use case... indeed, in the "individual roi-based" case, we know that a non-parametric test based on label-permutation of the input samples provides a valid solution which is straighforward to implement... (whereas at the "group-level", I don't have the same level of certainty ;) )

**Comment by bthirion** [comment]
Individual level means taking care of serial dependence etc, while at group level your data are essentially iid.

**Comment by bthirion** [comment]
Should we plan for a discussion on Discord. At 3 UTC ?

**Comment by SylvainTakerkart** [comment]
> Individual level means taking care of serial dependence etc, 

indeed you're right... but for the individual level, I was thinking about using as inputs the results of a first level GLM (with one regressor per trial), so that we get fairly iid data (the beta- or t-maps), which I believe is a fairly standard practice by now I think (but I might be wrong)

> while at group level your data are essentially iid.

at the group level, that means doing stats on first-level accuracies (that's what you're thinking about, right?), which is not trivial... (from what I know of the literature, there's no clearly valid solution that starts from first-level accuracies... but maybe I missed something?)

not clear what's most simple...

can't do 3UTC, sorry...

**Comment by bthirion** [comment]
When are you available to chat ?
